File: python3/day06/day06.py
answers_groups = open("input.txt").read().split("\n\n")


def part1():
    return sum(len(set(group.replace("\n", ""))) for group in answers_groups)


def part2():
    # OK this one looks weird I admit.
    # we create a set from each line (== each person's anwers) with map(set, ...), e.g. {'d', 'w', 't', 'b'}
    # *map unpacks the value of the map object to a list of sets
    # then we want to know the intersection of all person's answers in the group
    return sum(
        len(set.intersection(*map(set, group.splitlines()))) for group in answers_groups
    )


print(part1())
print(part2())

File: scala/src/main/scala/year2020/Day06.scala
package com.ochronus
package year2020

import utils.File


class Day06 {
  private val groups = File.readFileString("day06.input.txt").split("\n\n").map(_.linesIterator.toSeq)

  def part1(): Long = groups.map(_.reduce(_ ++ _).toSet.size).sum

  def part2(): Long = groups.map(group => group.map(_.toSet).reduce(_ & _).size).sum
}
